cmd/gator/validate/validate.go

```diff
@@ -0,0 +1,186 @@
+	Cmd.Flags().StringVarP(&flagOutput, flagNameOutput, "o", "", fmt.Sprintf("Output format.  One of: %s|%s.", stringJSON, stringYAML))
+
+	errLog = log.New(os.Stderr, "", 0)
+	outLog = log.New(os.Stdout, "", 0)
+}
+
+func run(cmd *cobra.Command, args []string) {
+	var unstrucs []*unstructured.Unstructured
+
+	// check if stdin has data
+	stdinfo, err := os.Stdin.Stat()
+	if err != nil {
+		errLog.Fatalf("getting info for stdout: %s", err)
+	}
+
+	// using stdin in combination with flags is not supported
+	if stdinfo.Size() > 0 && len(flagFilenames) > 0 {
+		errLog.Fatalf("stdin cannot be used in combination with %q flag", flagNameFilename)
+	}
+
+	// if no files specified, read from Stdin
+	switch {
+	case stdinfo.Size() > 0:
+		us, err := gator.ReadK8sResources(os.Stdin)
+		if err != nil {
+			errLog.Fatalf("reading from stdin: %s", err)
+		}
+		unstrucs = append(unstrucs, us...)
+	case len(flagFilenames) > 0:
+		us, err := readFiles(flagFilenames)
+		if err != nil {
+			errLog.Fatalf("reading from filenames: %s", err)
+		}
+		unstrucs = append(unstrucs, us...)
+	default:
+		errLog.Fatalf("no input data: must include data via either stdin or the %q flag", flagNameFilename)
+	}
+
+	responses, err := validate.Validate(unstrucs)
+	if err != nil {
+		errLog.Fatalf("auditing objects: %v\n", err)
+	}
+
+	results := responses.Results()
+
+	// TODO (https://github.com/open-policy-agent/gatekeeper/issues/1787): Add
+	// `-ojson` and `-oyaml` flags
+	switch flagOutput {
+	case stringJSON:
+		b, err := json.MarshalIndent(results, "", "    ")
+		if err != nil {
+			errLog.Fatalf("marshaling validation json results: %s", err)
+		}
+		outLog.Fatal(string(b))
+	case stringYAML:
+		b, err := yaml.Marshal(results)
+		if err != nil {
+			errLog.Fatalf("marshaling validation yaml results: %s", err)
+		}
+		outLog.Fatal(string(b))
+	default:
+		if len(results) > 0 {
+			for _, result := range results {
+				outLog.Printf("Message: %q", result.Msg)
+			}
+			os.Exit(1)
+		}
+	}
+}
+
+func readFiles(filenames []string) ([]*unstructured.Unstructured, error) {
+	var unstrucs []*unstructured.Unstructured
+
+	// normalize directories by listing their files
+	normalized, err := normalize(filenames)
+	if err != nil {
+		return nil, fmt.Errorf("normalizing: %w", err)
+	}
+
+	for _, filename := range normalized {
+		file, err := os.Open(filename)
+		if err != nil {
+			return nil, fmt.Errorf("opening file %q: %w", filename, err)
+		}
+
+		us, err := gator.ReadK8sResources(bufio.NewReader(file))
+		if err != nil {
+			return nil, fmt.Errorf("reading from file %q: %w", filename, err)
+		}
+		file.Close()
+
+		unstrucs = append(unstrucs, us...)
+	}
+
+	return unstrucs, nil
+}
+
+func normalize(filenames []string) ([]string, error) {
+	var output []string
+
+	for _, filename := range filenames {
+		err := filepath.Walk(filename, func(path string, info os.FileInfo, err error) error {
```

Comment:
nit: Since this anonymous function is longer than ten lines, it makes the logic of `normalize` gets a bit hard to follow. It's not a clear-cut case, but borderline for me so I'm making this a nit.